src/crypto/ecdsa.rs

```diff
@@ -0,0 +1,137 @@
+                },
+                PlainSecretParams::ECDSA(Mpi::from_raw_slice(secret.to_be_bytes().as_slice())),
+            ))
+        }
+
+        ECCCurve::P384 => {
+            let secret = p384::SecretKey::random(rng);
+            let public = secret.public_key();
+            Ok((
+                PublicParams::ECDSA {
+                    curve,
+                    p: Mpi::from_raw_slice(public.to_encoded_point(false).as_bytes()),
+                },
+                PlainSecretParams::ECDSA(Mpi::from_raw_slice(secret.to_be_bytes().as_slice())),
+            ))
+        }
+
+        _ => unsupported_err!("curve {:?} for ECDSA", curve),
+    }
+}
+
+/// Verify an ECDSA signature.
+pub fn verify(
+    curve: &ECCCurve,
+    p: &[u8],
+    _hash: HashAlgorithm,
+    hashed: &[u8],
+    sig: &[Mpi],
+) -> Result<()> {
+    match *curve {
+        ECCCurve::P256 => {
+            let flen = 32;
+            ensure_eq!(sig.len(), 2);
+
+            let r = sig[0].as_bytes();
+            let s = sig[1].as_bytes();
+
+            ensure!(r.len() <= flen, "invalid R (len)");
+            ensure!(s.len() <= flen, "invalid S (len)");
+            ensure_eq!(p.len(), 2 * flen + 1, "invalid P (len)");
+            ensure_eq!(p[0], 0x04, "invalid P (prefix)");
+
+            let pk = p256::ecdsa::VerifyingKey::from_sec1_bytes(p)?;
+            let mut sig_bytes = vec![0u8; 2 * flen];
+            //
+            // add padding if the values were encoded short
+            sig_bytes[(flen - r.len())..flen].copy_from_slice(r);
+            sig_bytes[flen + (flen - s.len())..].copy_from_slice(s);
+
+            let sig = p256::ecdsa::Signature::from_bytes(&sig_bytes)?;
+
+            pk.verify_prehash(hashed, &sig)?;
+
+            Ok(())
+        }
+        ECCCurve::P384 => {
+            let flen = 48;
+            ensure_eq!(sig.len(), 2);
+
+            let r = sig[0].as_bytes();
+            let s = sig[1].as_bytes();
+
+            ensure!(r.len() <= flen, "invalid R (len)");
+            ensure!(s.len() <= flen, "invalid S (len)");
+            ensure_eq!(p.len(), 2 * flen + 1, "invalid P (len)");
+            ensure_eq!(p[0], 0x04, "invalid P (prefix)");
+
+            let pk = p384::ecdsa::VerifyingKey::from_sec1_bytes(p)?;
+            let mut sig_bytes = vec![0u8; 2 * flen];
+            //
+            // add padding if the values were encoded short
+            sig_bytes[(flen - r.len())..flen].copy_from_slice(r);
+            sig_bytes[flen + (flen - s.len())..].copy_from_slice(s);
+
+            let sig = p384::ecdsa::Signature::from_bytes(&sig_bytes)?;
+
+            pk.verify_prehash(hashed, &sig)?;
+
+            Ok(())
+        }
+        _ => unsupported_err!("curve {:?} for ECDSA", curve.to_string()),
+    }
+}
+
+/// Sign using ECDSA
+pub fn sign(
+    curve: &ECCCurve,
+    secret_key: &ECDSASecretKey,
+    _hash: HashAlgorithm,
+    digest: &[u8],
+) -> Result<Vec<Vec<u8>>> {
```

Comment:
could be a tuple `Result<(Vec<u8>, Vec<u8>)>` instead of the additional vec allocation

Reply:
I followed other PK systems here. `crate::crypto::rsa::sign()` and `crate::crypto::eddsa::sign()` return Vec of Vec here. See `src/packet/secret_key_macro.rs`.

Reply:
fair, I should probably clean this up at some point 😅